lang/csharp/src/apache/test/Reflect/TestUnion.cs

```diff
@@ -0,0 +1,184 @@
+        public void NullableTest()
+        {
+            var nullableSchema = @"
+            [
+                ""null"",
+                { ""type"" : ""record"", ""name"" : ""Dervied2"", ""fields"" :
+                    [
+                        { ""name"" : ""A"", ""type"" : ""string""},
+                        { ""name"" : ""C"", ""type"" : ""double""}
+                    ]
+                },
+
+            ]
+            ";
+            var schema = Schema.Parse(nullableSchema);
+            var derived2write = new Derived2() { A = "derived2", C = 3.14 };
+
+            var writer = new ReflectWriter<Derived2>(schema);
+            var reader = new ReflectReader<Derived2>(schema, schema);
+
+            using (var stream = new MemoryStream(256))
+            {
+                var encoder = new BinaryEncoder(stream);
+                writer.Write(derived2write, encoder);
+                stream.Seek(0, SeekOrigin.Begin);
+
+                var decoder = new BinaryDecoder(stream);
+                var derived2read = reader.Read(decoder);
+                Assert.AreEqual(derived2read.A, derived2write.A);
+                Assert.AreEqual(derived2read.C, derived2write.C);
+            }
+        }
+
+        [TestCase]
+        public void HeterogeneousTest()
+        {
+            var heterogeneousSchema = @"
+            [
+                ""string"",
+                ""null"",
+                { ""type"" : ""record"", ""name"" : ""Dervied2"", ""fields"" :
+                    [
+                        { ""name"" : ""A"", ""type"" : ""string""},
+                        { ""name"" : ""C"", ""type"" : ""double""}
+                    ]
+                },
+
+            ]
+            ";
+            var schema = Schema.Parse(heterogeneousSchema);
+            var derived2write = new Derived2() { A = "derived2", C = 3.14 };
+
+            // union types (except for [null, type]) need to be manually registered
+            var unionSchema = schema as UnionSchema;
+            var cache = new ClassCache();
+            cache.LoadClassCache(typeof(Derived2), unionSchema[2]);
+
+            var writer = new ReflectWriter<object>(schema, cache);
+            var reader = new ReflectReader<object>(schema, schema, cache);
+
+            using (var stream = new MemoryStream(256))
+            {
+                var encoder = new BinaryEncoder(stream);
+                writer.Write(derived2write, encoder);
+                writer.Write("string value", encoder);
+                stream.Seek(0, SeekOrigin.Begin);
+
+                var decoder = new BinaryDecoder(stream);
+                var derived2read = (Derived2)reader.Read(decoder);
+                var stringRead = (string)reader.Read(decoder);
+                Assert.AreEqual(derived2read.A, derived2write.A);
+                Assert.AreEqual(derived2read.C, derived2write.C);
+                Assert.AreEqual(stringRead, "string value");
+            }
+        }
+    }
+
```

Comment:
Could you add another test that confirms that we throw an exception when you fail to manually register types?

```suggestion

        /// <summary>
        /// If you fail to manually register types within a union that has more than one non-null
        /// schema, creating a <see cref="ReflectWriter{T}"/> throws an exception.
        /// </summary>
        [TestCase]
        public void ThrowsIfClassesNotLoadedTest()
        {
            var schema = Schema.Parse(BaseClassSchema);
            var cache = new ClassCache();

            Assert.Throws<AvroException>(() => new ReflectWriter<BaseClass>(schema, cache));
        }
```

This suggestion reuses the `baseClassSchema` from the `BaseClassTest`, but I broke it out into a `const` in `TestUnion`:

```c#
        public const string BaseClassSchema = @"
        [
            { ""type"" : ""record"", ""name"" : ""Dervied1"", ""fields"" :
                [
                    { ""name"" : ""A"", ""type"" : ""string""},
                    { ""name"" : ""B"", ""type"" : ""int""}
                ]
            },
            { ""type"" : ""record"", ""name"" : ""Dervied2"", ""fields"" :
                [
                    { ""name"" : ""A"", ""type"" : ""string""},
                    { ""name"" : ""C"", ""type"" : ""double""}
                ]
            },

        ]
        ";
```